### ISSUE TITLE
Build fails on msys/mingw64/gcc8.1/windows 10 on AMD EPYC platform.

### ISSUE THREAD

**Issue by cosstas** [flame/blis#278]
Following the solution on issue #277, I encountered the following issues with blis commit 4bbb454 (that appear to be more serious and related):

`./configure auto` fails but explicitly defining architecture works, i.e. `./configure zen`:

> $ ./configure auto
configure: detected MINGW64_NT-10.0 kernel version 2.11.1(0.329/5/3).
configure: python interpeter search list is: python python3 python2.
configure: using 'python' python interpreter.
configure: found python version 3.7.1 (maj: 3, min: 7, rev: 1).
configure: python 3.7.1 appears to be supported.
configure: C compiler search list is: gcc clang cc.
configure: using 'gcc' C compiler.
configure: C++ compiler search list is: g++ clang++ c++.
configure: using 'g++' C++ compiler (for sandbox only).
configure: found gcc version 8.1.0 (maj: 8, min: 1, rev: 0).
configure: checking for blacklisted configurations due to gcc 8.1.0.
configure: found assembler ('as') version 2.30 (maj: 2, min: 30, rev: ).
configure: checking for blacklisted configurations due to as 2.30.
configure: reading configuration registry...done.
configure: determining default version string.
configure: could not find '.git' directory; using unmodified version file.
configure: starting configuration of BLIS 0.5.0.
configure: configuring with official version string.
configure: found shared library .so version '1.0.0'.
configure:   .so major version: 1
configure:   .so minor.build version: 0.0
configure: automatic configuration requested.
C:\msys64\tmp\ccoqSEGQ.o:bli_cpuid.c:(.text+0x0): multiple definition of `localtime_s'
C:\msys64\tmp\ccHSmelK.o:bli_arch.c:(.text+0x0): first defined here
C:\msys64\tmp\ccoqSEGQ.o:bli_cpuid.c:(.text+0x26): multiple definition of `gmtime_s'
C:\msys64\tmp\ccHSmelK.o:bli_arch.c:(.text+0x26): first defined here
C:\msys64\tmp\ccoqSEGQ.o:bli_cpuid.c:(.text+0x4c): multiple definition of `ctime_s'
C:\msys64\tmp\ccHSmelK.o:bli_arch.c:(.text+0x4c): first defined here
C:\msys64\tmp\ccwDUsns.o:config_detect.c:(.text+0x0): multiple definition of `localtime_s'
C:\msys64\tmp\ccHSmelK.o:bli_arch.c:(.text+0x0): first defined here
C:\msys64\tmp\ccwDUsns.o:config_detect.c:(.text+0x26): multiple definition of `gmtime_s'
C:\msys64\tmp\ccHSmelK.o:bli_arch.c:(.text+0x26): first defined here
C:\msys64\tmp\ccwDUsns.o:config_detect.c:(.text+0x4c): multiple definition of `ctime_s'
C:\msys64\tmp\ccHSmelK.o:bli_arch.c:(.text+0x4c): first defined here
collect2.exe: error: ld returned 1 exit status
./configure: line 1027: ./auto-detect.x: No such file or directory
configure: hardware detection driver returned ''.
configure: checking configuration against contents of 'config_registry'.
configure: 'auto-detected configuration '' is NOT registered!
configure:
configure: *** Cannot continue with unregistered configuration ''. ***
configure:

After configuring with './configure zen', 'make' fails again with multiple definition errors of time related functions. I attach the full build log:
[msys2_build_log.txt](https://github.com/flame/blis/files/2579605/msys2_build_log.txt)

**Comment by fgvanzee** [comment]
@cosstas Thanks for this follow-up report. I think @isuruf is most qualified to respond, so we'll wait for him to give his assessment.

**Comment by isuruf** [comment]
I've no idea why that is. Can you try disabling threads?

**Comment by cosstas** [comment]
It appears that I can't? Adding the ` --disable-threading` flag stops configuration step after checking library version and prints the help message:

> $ ./configure zen --disable-threading
configure: detected MINGW64_NT-10.0 kernel version 2.11.1(0.329/5/3).
configure: python interpeter search list is: python python3 python2.
configure: using 'python' python interpreter.
configure: found python version 3.7.1 (maj: 3, min: 7, rev: 1).
configure: python 3.7.1 appears to be supported.
configure: C compiler search list is: gcc clang cc.
configure: using 'gcc' C compiler.
configure: C++ compiler search list is: g++ clang++ c++.
configure: using 'g++' C++ compiler (for sandbox only).
configure: found gcc version 8.1.0 (maj: 8, min: 1, rev: 0).
configure: checking for blacklisted configurations due to gcc 8.1.0.
configure: found assembler ('as') version 2.30 (maj: 2, min: 30, rev: ).
configure: checking for blacklisted configurations due to as 2.30.
configure: reading configuration registry...done.
configure: determining default version string.
configure: could not find '.git' directory; using unmodified version file.
configure: starting configuration of BLIS 0.5.0.
configure: configuring with official version string.
configure: found shared library .so version '1.0.0'.
configure:   .so major version: 1
configure:   .so minor.build version: 0.0

 >configure (BLIS 0.5.0)

 >Field G. Van Zee

 >Configure BLIS's build system for compilation using a specified
 configuration directory.

The rest of the help message omitted.

**Comment by fgvanzee** [comment]
> $ ./configure zen --disable-threading

Based on the output above, you forgot the configuration target (e.g. `auto` or `zen`, for example).

**Comment by fgvanzee** [comment]
Nevermind, I somehow missed the `zen` upon first reading.

**Comment by fgvanzee** [comment]
Oh, here is the likely reason: you must give all options *before* the configuration target, as in:
```
 ./configure --disable-threading zen
```
That's why my eyes didn't catch it at first! :)

**Comment by cosstas** [comment]
No, the problem was that zen should follow after the --disable-threading flag. It is getting late here... I configured with the --disable-threading, but it fails the make step pretty much the same way...

**Comment by isuruf** [comment]
@cosstas, do you get multiple definition errors?

**Comment by devinamatthews** [comment]
All options have to go before `zen`.

**Comment by cosstas** [comment]
@isuruf Yes, here are the first error messages:
> Archiving lib/zen/libblis.a
Dynamically linking lib/zen/libblis.lib
obj/zen/kernels/zen/1/bli_amaxv_zen_int.o:bli_amaxv_zen_int.c:(.text+0x0): multiple definition of `_wctime'
obj/zen/config/zen/bli_cntx_init_zen.o:bli_cntx_init_zen.c:(.text+0x0): first defined here
obj/zen/kernels/zen/1/bli_amaxv_zen_int.o:bli_amaxv_zen_int.c:(.text+0x5): multiple definition of `_wctime_s'
obj/zen/config/zen/bli_cntx_init_zen.o:bli_cntx_init_zen.c:(.text+0x10): first defined here
obj/zen/kernels/zen/1/bli_amaxv_zen_int.o:bli_amaxv_zen_int.c:(.text+0x10): multiple definition of `difftime'
obj/zen/config/zen/bli_cntx_init_zen.o:bli_cntx_init_zen.c:(.text+0x20): first defined here
obj/zen/kernels/zen/1/bli_amaxv_zen_int.o:bli_amaxv_zen_int.c:(.text+0x15): multiple definition of `ctime'
obj/zen/config/zen/bli_cntx_init_zen.o:bli_cntx_init_zen.c:(.text+0x30): first defined here
obj/zen/kernels/zen/1/bli_amaxv_zen_int.o:bli_amaxv_zen_int.c:(.text+0x20): multiple definition of `gmtime'

**Comment by cosstas** [comment]
And similarly the auto configuration fails as before:

$ ./configure --disable-threading auto
configure: detected MINGW64_NT-10.0 kernel version 2.11.1(0.329/5/3).
configure: python interpeter search list is: python python3 python2.
configure: using 'python' python interpreter.
configure: found python version 3.7.1 (maj: 3, min: 7, rev: 1).
configure: python 3.7.1 appears to be supported.
configure: C compiler search list is: gcc clang cc.
configure: using 'gcc' C compiler.
configure: C++ compiler search list is: g++ clang++ c++.
configure: using 'g++' C++ compiler (for sandbox only).
configure: found gcc version 8.1.0 (maj: 8, min: 1, rev: 0).
configure: checking for blacklisted configurations due to gcc 8.1.0.
configure: found assembler ('as') version 2.30 (maj: 2, min: 30, rev: ).
configure: checking for blacklisted configurations due to as 2.30.
configure: reading configuration registry...done.
configure: determining default version string.
configure: could not find '.git' directory; using unmodified version file.
configure: starting configuration of BLIS 0.5.0.
configure: configuring with official version string.
configure: found shared library .so version '1.0.0'.
configure:   .so major version: 1
configure:   .so minor.build version: 0.0
configure: automatic configuration requested.
C:\msys64\tmp\ccPVB8Fz.o:bli_cpuid.c:(.text+0x0): multiple definition of `localtime_s'
C:\msys64\tmp\ccZgwIMV.o:bli_arch.c:(.text+0x0): first defined here
C:\msys64\tmp\ccPVB8Fz.o:bli_cpuid.c:(.text+0x26): multiple definition of `gmtime_s'
C:\msys64\tmp\ccZgwIMV.o:bli_arch.c:(.text+0x26): first defined here
C:\msys64\tmp\ccPVB8Fz.o:bli_cpuid.c:(.text+0x4c): multiple definition of `ctime_s'
C:\msys64\tmp\ccZgwIMV.o:bli_arch.c:(.text+0x4c): first defined here
C:\msys64\tmp\ccYLsQGC.o:config_detect.c:(.text+0x0): multiple definition of `localtime_s'
C:\msys64\tmp\ccZgwIMV.o:bli_arch.c:(.text+0x0): first defined here
C:\msys64\tmp\ccYLsQGC.o:config_detect.c:(.text+0x26): multiple definition of `gmtime_s'
C:\msys64\tmp\ccZgwIMV.o:bli_arch.c:(.text+0x26): first defined here
C:\msys64\tmp\ccYLsQGC.o:config_detect.c:(.text+0x4c): multiple definition of `ctime_s'
C:\msys64\tmp\ccZgwIMV.o:bli_arch.c:(.text+0x4c): first defined here
collect2.exe: error: ld returned 1 exit status
./configure: line 1027: ./auto-detect.x: No such file or directory
configure: hardware detection driver returned ''.
configure: checking configuration against contents of 'config_registry'.
configure: 'auto-detected configuration '' is NOT registered!
configure:
configure: *** Cannot continue with unregistered configuration ''. ***
configure:

**Comment by fgvanzee** [comment]
It's pretty clear that the auto-detection binary is not being built (due to the multiple definitions errors), meaning it never gets a configuration name back from the binary, hence the `configure` error message at the end. But it's not clear at all where those extra definitions are coming from.

**Comment by isuruf** [comment]
Those symbols are not from blis and I don't know why they are being compiled into the 3 object files. Maybe the toolchain you are using is broken.

**Comment by cosstas** [comment]
I could/will try different toolchains and versions next morning. And anything else you may come up with in the meantime...

**Comment by cosstas** [comment]
Here's the update:
- Tried mingw64/gcc 7.3.1 and it fails just like gcc8.1.0.
- Went back to 8.1.0 and added `--disable-shared` flag. blis compiled successfully and libblis.a created.
- Tried to build lapack-3.8.0 and link against libblis.a and then got the same multiple definition errors related to time.
- Reinstalled MSYS2 with native gcc8.2.0 (the previous was an MSYS2 installation with MINGW-W64 toolchains on top). Made no difference at all at building blis. Again multiple definition errors.
So back to where I started. A search on the internet pointed to the different treatment of time function definitions (related to time.h) between linux and windows but that's up to where my abilities reach. 
If there is anything else I can do on my side, please say so.

**Comment by isuruf** [comment]
Can you try doing,
```
export CFLAGS="-D_MT -D_DLL"
```

rerun configure and build again?

**Comment by cosstas** [comment]
Did so, configure took the flags into account:
> configure: detected preset CFLAGS; prepending:
configure:   -D_MT -D_DLL
It made no difference in the build, it failed the same way.

**Comment by cosstas** [comment]
As mentioned in #252, I can eliminate the errors at the end related to cblas, e.g.

>  Cannot export cblas_caxpy: symbol not defined

by enabling cblas with `--enable-cblas`. Only the multiple definitions remain. @isuruf But I guess you already knew that.

**Comment by isuruf** [comment]
I looked at the headers,
```
mingw64/x86_64-w64-mingw32/include/time.h:__forceinline errno_t __cdecl localtime_s(struct tm *_Tm,const time_t *_Time) { return _localtime32_s(_Tm,_Time); }
mingw64/x86_64-w64-mingw32/include/time.h:__forceinline errno_t __cdecl gmtime_s(struct tm *_Tm, const time_t *_Time)   { return _gmtime32_s(_Tm, _Time); }
mingw64/x86_64-w64-mingw32/include/time.h:__forceinline errno_t __cdecl ctime_s(char *_Buf,size_t _SizeInBytes,const time_t *_Time) { return _ctime32_s(_Buf,_SizeInBytes,_Time); }
mingw64/x86_64-w64-mingw32/include/time.h:__forceinline errno_t __cdecl localtime_s(struct tm *_Tm,const time_t *_Time) { return _localtime64_s(_Tm,_Time); }
mingw64/x86_64-w64-mingw32/include/time.h:__forceinline errno_t __cdecl gmtime_s(struct tm *_Tm, const time_t *_Time) { return _gmtime64_s(_Tm, _Time); }
mingw64/x86_64-w64-mingw32/include/time.h:__forceinline errno_t __cdecl ctime_s(char *_Buf,size_t _SizeInBytes,const time_t *_Time) { return _ctime64_s(_Buf,_SizeInBytes,_Time); }
mingw64/x86_64-w64-mingw32/include/time.h:__forceinline struct tm *__cdecl localtime_r(const time_t *_Time, struct tm *_Tm) {
mingw64/x86_64-w64-mingw32/include/time.h:__forceinline struct tm *__cdecl gmtime_r(const time_t *_Time, struct tm *_Tm) {
mingw64/x86_64-w64-mingw32/include/time.h:__forceinline char *__cdecl ctime_r(const time_t *_Time, char *_Str) {
mingw64/x86_64-w64-mingw32/include/time.h:__forceinline char *__cdecl asctime_r(const struct tm *_Tm, char * _Str) {
```

They are inline functions and since their declarations are the same, the linker should ignore them. My suggestion is to try to reproduce this without blis, by compiling two files with time.h included. If the error can be reproduced in an example without blis, you'll have to post in the mingw list.

Btw, have you tried clang? Do you need to use mingw?

**Comment by cosstas** [comment]
> They are inline functions and since their declarations are the same, the linker should ignore them. My suggestion is to try to reproduce this without blis, by compiling two files with time.h included.

This is well beyond my current scope of understanding and won't be able to follow up.

> Btw, have you tried clang? Do you need to use mingw?

The main scientific code I am using is written in FORTRAN so clang won't be of much use there.

**Comment by cosstas** [comment]
I have tested the fixes proposed in #283 and it appears that now blis compiles successfully. Also `./configure auto` now works as well. I will now test how the built library links with lapack etc and report.
Many thanks for your efforts!

**Comment by cosstas** [comment]
OK, all BLIS and BLAS tests from the testsuite pass on my platform, so I consider it case closed...

**Comment by fgvanzee** [comment]
Glad @isuruf was able to accommodate your needs, @cosstas. Please keep in touch.

**Comment by cosstas** [comment]
For the sake of completeness, I confirm that undefined symbols errors were also addressed with #283 so the `--enable-cblas` workaround is not required anymore. Great work by @isuruf!
